crates/rattler_conda_types/src/explicit_environment_spec.rs

```diff
@@ -0,0 +1,291 @@
+    pub fn package_archive_hash(
+        &self,
+    ) -> Result<Option<PackageArchiveHash>, ParsePackageArchiveHashError> {
+        self.url
+            .fragment()
+            .map_or(Ok(None), |s| PackageArchiveHash::from_str(s).map(Some))
+    }
+}
+
+impl From<Url> for ExplicitEnvironmentEntry {
+    fn from(url: Url) -> Self {
+        ExplicitEnvironmentEntry { url }
+    }
+}
+
+impl From<ExplicitEnvironmentEntry> for Url {
+    fn from(entry: ExplicitEnvironmentEntry) -> Self {
+        entry.url
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ParseExplicitEnvironmentSpecError {
+    #[error("the @EXPLICIT tag is missing")]
+    MissingExplicitTag,
+
+    #[error("failed to parse url '{0}'")]
+    InvalidUrl(String, #[source] url::ParseError),
+
+    #[error(transparent)]
+    InvalidPlatform(#[from] ParsePlatformError),
+
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+}
+
+impl ExplicitEnvironmentSpec {
+    /// Parses an explicit environment file from a reader.
+    pub fn from_reader(mut reader: impl Read) -> Result<Self, ParseExplicitEnvironmentSpecError> {
+        let mut str = String::new();
+        reader.read_to_string(&mut str)?;
+        Self::from_str(&str)
+    }
+
+    /// Parses an explicit environment file from a file.
+    pub fn from_path(path: &Path) -> Result<Self, ParseExplicitEnvironmentSpecError> {
+        Self::from_reader(File::open(path)?)
+    }
+}
+
+impl FromStr for ExplicitEnvironmentSpec {
+    type Err = ParseExplicitEnvironmentSpecError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut platform = None;
+        let mut is_explicit = false;
+        let mut packages = Vec::new();
+        for line in s.lines() {
+            // Skip lines starting with a #
+            if let Some(comment_line) = line.strip_prefix('#') {
+                // Unless that comment line is `# platform: `. Because then we're interested in the
+                // platform specifier.
+                if let Some(platform_str) = comment_line.trim_start().strip_prefix("platform:") {
+                    platform = Some(Platform::from_str(platform_str.trim())?);
+                }
+            } else if line.trim() == "@EXPLICIT" {
+                is_explicit = true
+            } else if !is_explicit {
+                return Err(ParseExplicitEnvironmentSpecError::MissingExplicitTag);
+            } else {
+                // Parse the line as an explicit URL
+                packages.push(
+                    Url::parse(line.trim())
+                        .map_err(|e| {
+                            ParseExplicitEnvironmentSpecError::InvalidUrl(line.trim().to_owned(), e)
+                        })?
+                        .into(),
+                );
+            }
+        }
+
+        if !is_explicit {
+            return Err(ParseExplicitEnvironmentSpecError::MissingExplicitTag);
```

Comment:
Can this ever be the case? Since you cover it already when iterating over the lines.